### DISCUSSION TITLE
Add configuration option to disable HTTP-only flag for refresh token cookie

### DISCUSSION THREAD

**Discussion by Cel51** [directus/directus#21715] [Feature Requests]
### Summary

This feature request proposes adding a configuration option in Directus to allow disabling the HTTP-only flag for the refresh token cookie. By providing this option, Directus can offer more flexibility for developers integrating Directus with client-side frameworks like Vue.js, where the refresh token needs to be accessed on the client-side for subsequent API requests.

### Basic Example

```.env
# default to true
REFRESH_TOKEN_HTTPONLY=false
```

### Motivation

The motivation behind this feature request is to enhance the integration capabilities of Directus with client-side frameworks, particularly Vue.js. Currently, Directus sets the "directus_refresh_token" cookie as HTTP-only by default, preventing client-side access to the refresh token. This poses a challenge when building applications that rely on client-side token management and authentication flows.

By providing a configuration option to disable the HTTP-only flag for the refresh token cookie, developers will have the flexibility to handle authentication and token management according to their application's requirements. This will enable seamless integration with client-side frameworks and allow for more streamlined login flows without the need for server-side routes to handle token refresh.

### Detailed Design

The proposed feature involves adding a new configuration option in Directus named REFRESH_TOKEN_HTTPONLY (or a similar name) that determines whether the refresh token cookie should be set as HTTP-only or not. The configuration option will be a boolean value, with the default value set to true to maintain the current behavior and ensure backward compatibility.

When REFRESH_TOKEN_HTTPONLY is set to false, Directus will not set the HTTP-only flag for the refresh token cookie, allowing client-side access to the refresh token. This will enable developers to retrieve the refresh token on the client-side and use it for subsequent API requests, such as refreshing the access token when it expires.

The implementation of this feature will require modifying the code responsible for setting the refresh token cookie in Directus. The code will need to check the value of the refreshTokenHttpOnly configuration option and conditionally set the HTTP-only flag based on its value.

### Must Have:
- Add a new configuration option REFRESH_TOKEN_HTTPONLY to control the HTTP-only flag for the refresh token cookie.
- Set the default value of REFRESH_TOKEN_HTTPONLY to true to maintain backward compatibility.
- Modify the code responsible for setting the refresh token cookie to conditionally set the HTTP-only flag based on the REFRESH_TOKEN_HTTPONLY value.
- Information about security risk endorsed by disabling this functionality

### Should Have:
- Update the Directus documentation to include information about the new configuration option and its usage.

### Could Have:
- Provide examples and guidance on how to integrate Directus with client-side frameworks using the new configuration option.

### Drawbacks

The main drawback of implementing this feature is the potential security implications of allowing client-side access to the refresh token. The HTTP-only flag is a security best practice that prevents client-side JavaScript from accessing sensitive cookies, reducing the risk of cross-site scripting (XSS) attacks. By providing an option to disable the HTTP-only flag, there is a risk that developers may inadvertently expose the refresh token to client-side vulnerabilities if not handled properly.

However, this risk can be mitigated by providing clear documentation and guidelines on securely handling the refresh token on the client-side. Developers should be encouraged to follow best practices, such as storing the refresh token securely and protecting against XSS attacks.

### Alternatives

An alternative approach to this feature request would be to explore other mechanisms for handling authentication and token management in client-side frameworks. For example, using server-side routes to handle token refresh and providing APIs for client-side frameworks to interact with Directus authentication.

However, this alternative may require more complex server-side setup and may not provide the same level of flexibility and seamless integration as the proposed configuration option.

### Adoption Strategy

The adoption of this feature should be relatively straightforward for existing Directus developers. Since the proposed configuration option will have a default value that maintains the current behavior, existing applications will not be affected unless explicitly configured to disable the HTTP-only flag.

For developers who choose to use the new configuration option, they will need to update their Directus configuration file and handle the refresh token accordingly on the client-side. The Directus documentation should provide clear instructions

**Comment by rijkvanzanten** [comment]
> However, this risk can be mitigated by providing clear documentation and guidelines on securely handling the refresh token on the client-side. Developers should be encouraged to follow best practices, such as storing the refresh token securely and protecting against XSS attacks.

The way to do this is by using `httpOnly` cookies. As soon as you allow client JS to access the token itself, any javascript, including third party JS like analytics scripts or browser extensions can access the token.

**Reply by rijkvanzanten** [reply]
And if you really want the refresh token in insecurely in the the client JS, you could use the `json` response mode in login https://docs.directus.io/reference/authentication.html#login, but I'd highly recommend against doing so
